[SPARK-39301][SQL][PYTHON] Leverage LocalRelation and respect Arrow batch size in createDataFrame with Arrow optimization

HyukjinKwon opened the pull request:
### What changes were proposed in this pull request?

This PR proposes to use `LocalRelation` instead of `LogicalRDD` when creating a (small) DataFrame with Arrow optimization, which passes the data as a local data in the driver side (which is consistent with Scala code path).

Namely:

```python
import pandas as pd
spark.conf.set("spark.sql.execution.arrow.pyspark.enabled", True)
spark.createDataFrame(pd.DataFrame({'a': [1, 2, 3, 4]})).explain(True)
```

Before

```
== Parsed Logical Plan ==
LogicalRDD [a#0L], false

== Analyzed Logical Plan ==
a: bigint
LogicalRDD [a#0L], false

== Optimized Logical Plan ==
LogicalRDD [a#0L], false

== Physical Plan ==
*(1) Scan ExistingRDD arrow[a#0L]
```

After

```
== Parsed Logical Plan ==
LocalRelation [a#0L]

== Analyzed Logical Plan ==
a: bigint
LocalRelation [a#0L]

== Optimized Logical Plan ==
LocalRelation [a#0L]

== Physical Plan ==
LocalTableScan [a#0L]
```

This is controlled by a new configuration `spark.sql.execution.arrow.localRelationThreshold` defaulting to 48MB. This default was picked by benchmark I ran below.

In addition, this PR also fixes `createDataFrame` to respect `spark.sql.execution.arrow.maxRecordsPerBatch` configuration when creating Arrow bathes. Previously, we divided the input pandas DataFrame by the default partition number which forced users to set `spark.rpc.message.maxSize` when the input pandas DataFrame is too large. See the benchmark performed below.

### Why are the changes needed?

We have some nice optimization for `LocalRelation` (e.g., `ConvertToLocalRelation`). For example, the stats are fully known when you use `LocalRelation`. With `LogicalRDD`, many optimizations cannot be applied. Even in some cases (e.g., `executeCollect`), we can avoid creating `RDD`s too.

For respecting `spark.sql.execution.arrow.maxRecordsPerBatch`, 1. we can avoid forcing users to set `spark.rpc.message.maxSize`, and 2. I believe the configuration is supposed to be respected for all code path that creates Arrow batches if possible.

### Does this PR introduce _any_ user-facing change?

No, it is an optimization. The number of partitions can be different, but that should be internal.

### How was this patch tested?

- Manually tested.
- Added a unittest.
- I did two benchmark tests with 1 Driver & 4 Workers (i3.xlarge), see below.

#### Benchmark 1 (best cases)

```python
import time
import random
import string

import pandas as pd

names = [random.choice(list(string.ascii_lowercase)) for i in range(1000)]
ages = [random.randint(0, 100) for i in range(1000)]
l = list(zip(names, ages))
d = [{'name': a_name, 'age': an_age} for a_name, an_age in l]
pdf = pd.DataFrame({'name': names, 'age': ages})
spark.range(1).count()  # heat up

start = time.time()
for _ in range(100):
    _ = spark.createDataFrame(pdf)

end = time.time()
print(end - start)
```

Before

10.250491698582968

After

6.004616181055705

#### Benchmark 2 (worst cases)

```bash
curl -O https://eforexcel.com/wp/wp-content/uploads/2020/09/HR2m.zip
unzip HR2m.zip
```

```python
import pandas as pd
pdf = pd.read_csv("HR2m.csv")
pdf23 = pdf.iloc[:int(len(pdf)/32)]
pdf45 = pdf.iloc[:int(len(pdf)/16)]
pdf90 = pdf.iloc[:int(len(pdf)/8)]
pdf175 = pdf.iloc[:int(len(pdf)/4)]
pdf350 = pdf.iloc[:int(len(pdf)/2)]
pdf700 = pdf.iloc[:int(len(pdf))]
pdf2gb = pd.concat([pdf, pdf, pdf])
pdf5gb = pd.concat([pdf2gb, pdf2gb])

spark.createDataFrame(pdf23)._jdf.rdd().count()  # explicitly create RDD.
...
```

Before

23MB: 1.02 seconds
45MB: 1.69 seconds
90MB: 2.38 seconds
175MB: 3.19 seconds
350MB: 6.10 seconds
2GB: 43.21 seconds
5GB: X (threw an exception that says to set 'spark.rpc.message.size' higher)

After

23MB: 1.31 seconds (local collection is used)
45MB: 2.47 seconds (local collection is used)
90MB: 1.79 seconds
175MB: 3.22 seconds
350MB: 6.41 seconds
2GB: 47.12 seconds
5GB: 1.29 minutes

**NOTE** that the performance varies depending on network stability, and the numbers above are from second run (it's not the average).

HyukjinKwon committed:
Levearge LocalRelation in createDataFrame with Arrow optimization

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:
```diff
@@ -190,32 +191,30 @@ private[sql] object ArrowConverters {
+      TaskContext.get())
+    // Project it. Otherwise, the Arrow column vectors will be closed and released out.
+    val proj = UnsafeProjection.create(attrs, attrs)
+    Dataset.ofRows(session, LocalRelation(attrs, data.map(r => proj(r).copy()).toArray))
```
One downside of this approach is that, now the `data` will live in driver side. If the pandas DataFrame is too big, it can easily throw an exception from the driver side.

However, I think this is the same as the Scala side so I suspect this is fine. Maybe we can add a configuration to turn on and off but I don't feel strongly on that.

ueshin replied:
Also the conversion from byte array to `InternalRow` only happens in driver side that is not in parallel. There might be some threshold of the performance benefit.

ueshin replied:
Btw, if we don't handle it in parallel, we don't need to slice the input: https://github.com/apache/spark/blob/8053d1bce3904e15dfe494bfd7bc3f4e92c80037/python/pyspark/sql/pandas/conversion.py#L598-L606?

It could still reduce the memory consumption during the conversion, though.

JoshRosen replied:
I agree that the `ConvertToLocalRelation` optimization can greatly improve performance for small tables, especially because local relations' have an accurate implementation of `computeStats()` which allows us to benefit from other optimizations like auto broadcast joins.

I suppose it's possible that Arrow batches could be _much_ more compressible than UnsafeRows such that we experience a huge inflation in memory usage. For example, let's say I have an Arrow row batch containing a huge number of rows with all boolean fields: these booleans will be encoded space-efficiently in Arrow but will experience huge bloat when converted into UnsafeRows.

I don't have any sense of whether users are doing this today, though. _Probably_ most cases are small enough that this wouldn't be an issue. 

If we _are_ concerned about this, though, then here's some brainstorming on potential options:

- We could add a configuration flag so that users can opt-in to revert back to the old behavior. Users would have to somehow know/discover this flag after they hit performance problems, though.
- We might be able to come up with reasonably accurate heuristic estimates of the dataset's unssafe row size based on the schema and then choose which path to use based on the size threshold. Our new flag could be a size threshold flag (like autobroadcast join) rather than a boolean flag.
  - As a strawman, the threshold could be 10 MB (the default autobroadcast join threshold).
  - For the estimation, we can use `EstimationUtils.getSizePerRow(attributes) * numRows`. 
    - As an aside, I think that some of these estimates are potentially inaccurate because they're based on `DataType.defaultSize` which I think tends to underestimate UnsafeRow overheads for small primitive fields (e.g. it estimates the cost of a boolean as 1 byte when in reality it's 8 bytes in an UnsafeRow). That said, I think we should probably use the same estimation logic everywhere and should separately investigate if we want to re-tune that method for accuracy. Changes to that estimation would have wide-rippling effects on whether we autobroadcast join, so we'd need careful investigation to make sure we don't significantly regress existing queries if we change the estimation.

Do you think it's worth it to worry about this edge-case and add thresholding? Or is the risk/impact low enough that the extra complexity isn't justified?

ueshin replied:
+1 for adding the threshold flag.

viirya replied:
SGTM for a configuration flag.

HyukjinKwon commented:
cc @ueshin @viirya @BryanCutler FYI

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:
```diff
@@ -190,32 +191,30 @@ private[sql] object ArrowConverters {
+      DataType.fromJson(schemaString).asInstanceOf[StructType],
+      session.sessionState.conf.sessionLocalTimeZone,
+      TaskContext.get())
+    // Project it. Otherwise, the Arrow column vectors will be closed and released out.
```
```suggestion
    // Project/copy it. Otherwise, the Arrow column vectors will be closed and released out.
```

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/api/python/PythonSQLUtils.scala:
```diff
@@ -70,22 +70,22 @@ private[sql] object PythonSQLUtils extends Logging {
    */
-  def readArrowStreamFromFile(session: SparkSession, filename: String): JavaRDD[Array[Byte]] = {
-    ArrowConverters.readArrowStreamFromFile(session, filename)
+  def readArrowStreamFromFile(filename: String): Iterator[Array[Byte]] = {
```
Intentionally I used `Iterator` to avoid Py4J copies `Array` into Python driver side.

HyukjinKwon replied:
This is Scala Iterator so it doesn't get converted to Python iterator via Py4J.

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:
```diff
@@ -190,32 +191,30 @@ private[sql] object ArrowConverters {
-      ArrowConverters.fromBatchIterator(iter, schema, timeZoneId, context)
-    }
-    session.internalCreateDataFrame(rdd.setName("arrow"), schema)
+    val attrs = DataType.fromJson(schemaString).asInstanceOf[StructType].toAttributes
+    val data = ArrowConverters.fromBatchIterator(
+      arrowBatches,
+      DataType.fromJson(schemaString).asInstanceOf[StructType],
```
Maybe we can avoid calling `DataType.fromJson` twice if we store `schema` into a local variable and use a separate `attrs` variable? Like this: 

```suggestion
    val schema = DataType.fromJson(schemaString).asInstanceOf[StructType]
    val attrs = schema.toAttributes
    val data = ArrowConverters.fromBatchIterator(
      arrowBatches,
      schema,
```

JoshRosen commented on python/pyspark/sql/pandas/conversion.py:
```diff
@@ -613,16 +613,16 @@ def _create_from_pandas_with_arrow(
 
         # Create Spark DataFrame from Arrow stream file, using one batch per partition
-        jrdd = self._sc._serialize_to_jvm(arrow_data, ser, reader_func, create_RDD_server)
+        jiter = self._sc._serialize_to_jvm(arrow_data, ser, reader_func, create_RDD_server)
```
Here it looks like we have changed the type of `reader_func`: before it was a function from filename to `JavaRDD[Array[Byte]]`, whereas now it's a function from filename to `Iterator[Array[Byte]]`.

I think this breaks the current interface contract of `_serialize_to_jvm` in a way that will cause problems if IO encryption is enabled. According to [that method's docs](https://github.com/apache/spark/blob/dbde77856d2e51ff502a7fc1dba7f10316c2211b/python/pyspark/context.py#L677), `reader_func` is only used when IO encryption is disabled (which is the default configuration). If IO encryption is enabled then this codepath will use the `PythonParallelizeServer` codepath which will still produce a `JavaRDD[Array[Byte]]`.

To fix this, we might need to change the other pieces of this interface so that they all pass around `Iterator[Array[Byte]]` (in other words, modify the regular Python parallelize codepath so that its implementation is closer to what you have here).

Good catch about the performance risks of passing `Array[Array[Byte]]` back to Python via Py4J, BTW. Do you know if `ArrayBuffer` has the same issue? I'm wondering whether we could pass `ArrayBuffer[Array[Byte]]` if we want to ensure that the returned object is still considered `Iterable` from Python.

HyukjinKwon replied:
BTW, we have a test for encryption on and off https://github.com/apache/spark/blob/master/python/pyspark/sql/tests/test_arrow.py#L771-L774

HyukjinKwon commented:
Wow, thanks for reviews guys. Let me think a bit more and push some changes soon.

HyukjinKwon commented:
This PR is ready for a look now. TL;DR:
- I added one new configuration `spark.sql.execution.arrow.localRelationThreshold` as a threshold.
- I fixed the code path to respect `spark.sql.execution.arrow.maxRecordsPerBatch` when creating Arrow batches to avoid forcing users to set a larger `spark.rpc.message.maxSize`.

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2575,6 +2575,18 @@ object SQLConf {
+      .version("3.4.0")
+      .bytesConf(ByteUnit.BYTE)
+      .checkValue(_ >= 0, "This value must be equal to or greater than 0.")
+      .createWithDefaultString("64MB")
```
```suggestion
      .createWithDefaultString("48MB")
```

HyukjinKwon committed:
Test multiple partitions

HyukjinKwon commented:
Gentle ping for a review :-). I know it has some trade-off but I believe this addresses more common cases and benefit more users.

BryanCutler reviewed (commented):
Looks like it's a good optimization for smaller files, thanks @HyukjinKwon !

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -2575,6 +2575,18 @@ object SQLConf {
+      .version("3.4.0")
+      .bytesConf(ByteUnit.BYTE)
+      .checkValue(_ >= 0, "This value must be equal to or greater than 0.")
+      .createWithDefaultString("48MB")
```
It's all together ..so pretty small

HyukjinKwon commented on python/pyspark/sql/pandas/conversion.py:
```diff
@@ -596,7 +596,7 @@ def _create_from_pandas_with_arrow(
 
         # Slice the DataFrame to be batched
-        step = -(-len(pdf) // self.sparkContext.defaultParallelism)  # round int up
+        step = self._jconf.arrowMaxRecordsPerBatch()
```
Yeah, that's true... perf diff seems trivial in any event and seems it works around the `spark.rpc.message.maxSize` issue.

BryanCutler replied:
I believe it was like this to create the same number of partitions as when arrow is disabled, although that might have changed since. If the DataFrame is split with `arrowMaxRecordsPerBatch` and a user wanted to create a certain number of partitions, then would they have to look at the size of the input and then adjust `arrowMaxRecordsPerBatch` accordingly?

HyukjinKwon replied:
Yeah, that's true .. but I wonder if the default number of partitions is something we should consider given that it wasn't already configurable before, and `SparkSession.createDataFrame` does not expose the number of partitions too.

If they really need, users might want to create an RDD with an explicit parallelism .. we don't support this now though (see also https://github.com/apache/spark/pull/29719).

HyukjinKwon replied:
BTW, just to extra clarify, when the pandas DataFrame is small (lower than the threshold), the number of partitions remains same (configured by `spark.sql.leafNodeDefaultParallelism` that falls back to `sparkContext.defaultParallelism` if not set).

The number of partitions is only different when the input DataFrame is large.

HyukjinKwon commented:
cc @mengxr and @WeichenXu123 in case you guys have some comments.

HyukjinKwon commented:
Let me merge this in few days ... assuming that we're all good. Hopefully my benchmark is good enough.

HyukjinKwon commented:
Rebased

HyukjinKwon commented:
Let me get this in. It's the early stage of Spark 3.4 so should be good timing to merge such stuff.

Merged to master.
